dlib/dnn/layers.h

```diff
@@ -701,6 +701,149 @@ namespace dlib
 
 // ----------------------------------------------------------------------------------------
 
+    template <
+        long _repeat_r,
+        long _repeat_c
+        >
+    class block_upsample_
+    {
+        static_assert(_repeat_r >= 1, "The number of rows to repeat must be >= 1");
+        static_assert(_repeat_c >= 1, "The number of columns to repeat must be >= 1");
+    public:
+
+
+        block_upsample_(
+        )
+        {}
+
+        long repeat_r() const { return _repeat_r; }
+        long repeat_c() const { return _repeat_c; }
+
+        inline point map_input_to_output (
+            point p
+        ) const
+        {
+            p.x() = p.x()*repeat_c();
+            p.y() = p.y()*repeat_r();
+            return p;
+        }
+
+        inline point map_output_to_input (
+            point p
+        ) const
+        {
+            p.x() = p.x()/repeat_c();
+            p.y() = p.y()/repeat_r();
+            return p;
+        }
+
+        block_upsample_ (
+            const block_upsample_& item
+        )
+        {
+            // this->bu is non-copyable so we have to write our own copy to avoid trying to
```

Comment:
Should I implement a copy procedue here or let the compiler generate one if somebody calls it in code?